features/steps/packages.py

```diff
@@ -3,19 +3,91 @@
 from behave import then, when
 from hamcrest import assert_that, contains_string, matches_regexp
 
-from features.steps.shell import when_i_run_command
+from features.steps.shell import when_i_retry_run_command, when_i_run_command
 from features.util import SUT
 
 
-@when("I apt install `{package_names}`")
-@when("I apt install `{package_names}` on the `{machine_name}` machine")
-def when_i_apt_install(context, package_names, machine_name=SUT):
+@when("I remove unneeded packages from the system")
+def when_i_autoremove(context):
+    when_i_run_command(
+        context,
+        "apt -y autoremove",
+        "with sudo",
+    )
+
+
+@when("I perform a dist-upgrade")
+def when_i_dist_update(context):
     when_i_run_command(
+        context,
+        "apt dist-upgrade --assume-yes --allow-downgrades",
+        "with sudo",
+    )
+
+
+@then("I ensure apt update runs without errors")
+def then_i_check_apt_update(context):
+    when_i_run_command(
+        context,
+        "apt update",
+        "with sudo",
+    )
+
+
+@when("I update apt package lists")
+@when("I update apt package lists on the `{machine_name}` machine")
+def when_i_apt_update(context, machine_name=SUT):
+    when_i_retry_run_command(
+        context,
+        "apt update",
+        "with sudo",
+        machine_name=machine_name,
+        exit_codes="100",
+    )
+
+
+@when("I perform a {dry} run of {command} upgrade")
+@when("I upgrade all apt packages in the system")
+@when("I upgrade all apt packages in the system using {command}")
+@when("I upgrade all apt packages in the system including {phased} updates")
+def when_i_apt_upgrade(context, command="apt", dry="", phased=""):
+    cmd_list = [
+        command,
+        "upgrade",
+    ]
+
+    if dry == "dry":
+        cmd_list.append("--dry-run")
+    else:
+        cmd_list.insert(0, "DEBIAN_FRONTEND=noninteractive")
+        cmd_list.extend(
+            [
+                "-y",
+                "--allow-downgrades",
+                '-o Dpkg::Options::="--force-confdef"',
+                '-o Dpkg::Options::="--force-confold"',
+            ]
+        )
+    # It will be either phased or dry, never both
+    if phased == "phased":
+        cmd_list.append("-o APT::Get::Always-Include-Phased-Updates=true")
+
+    when_i_run_command(
+        context,
+        " ".join(cmd_list),
+        "with sudo",
+    )
+
+
+@when("I install `{package_names}`")
```

Comment:
I had it as "I apt install" to differentiate it from a snap install, do you think that is a useful differentiation?

Reply:
yes